File: frontend/todo-react-native/src/types/todo.ts
export type TodoStatus = 'todo' | 'doing' | 'done';

export interface Todo {
  _id: string;
  title: string;
  description?: string;
  dueDate?: string;
  status: TodoStatus;
  priority: number;
  createdAt: string;
  updatedAt: string;
  __v: number;
}

export interface TodoListResult {
  items: Todo[];
  total: number;
  page: number;
  limit: number;
}

export interface ListTodosParams {
  page?: number;
  limit?: number;
  status?: TodoStatus | 'all';
}

export interface CreateTodoPayload {
  title: string;
  description?: string;
  dueDate?: string;
  status?: TodoStatus;
  priority?: number;
}

export interface UpdateTodoPayload {
  title?: string;
  description?: string;
  dueDate?: string;
  status?: TodoStatus;
  priority?: number;
  __v?: number;
}

File: backend/todo-spring/src/main/java/com/todo/model/Todo.java
package com.todo.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.annotation.Version;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;

/**
 * MongoDB document for a Todo item.
 * {@code @Version} drives optimistic-locking: a concurrent write with a stale
 * version throws {@link org.springframework.dao.OptimisticLockingFailureException}.
 * The version is serialised as {@code __v} for API parity with the NestJS/Express backends.
 */
@Data
@NoArgsConstructor
@Document(collection = "todos")
@CompoundIndex(def = "{'status': 1, 'dueDate': 1}")
public class Todo {

    @Id
    private String id;

    private String title;

    private String description;

    private Instant dueDate;

    private TodoStatus status = TodoStatus.TODO;

    private int priority = 3;

    @CreatedDate
    private Instant createdAt;

    @LastModifiedDate
    private Instant updatedAt;

    @Version
    @JsonProperty("__v")
    private Long version;
}
